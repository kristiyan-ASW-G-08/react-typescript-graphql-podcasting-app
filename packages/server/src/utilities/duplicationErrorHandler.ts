import { NextFunction } from 'express';
import { Document } from 'mongoose';
import ValidationError from '@pod/common/source/types/ValidationError';
import { UserInputError } from 'apollo-server';
import logger from './logger';

const duplicationErrorHandler = (
  duplicationErrors: any,
  doc: Document,
  next: NextFunction,
): void => {
  if (duplicationErrors.errors) {
    const validationErrors: ValidationError[] = Object.values(
      duplicationErrors.errors,
    ).map(
      // @ts-ignore
      ({ value, path }: { value: string; path: string }): ValidationError => ({
        path,
        message: `${value} is already taken`,
      }),
    );
    throw new UserInputError('Validation Error', {
      validationErrors,
    });
  } else {
    next();
  }
};
export default duplicationErrorHandler;
