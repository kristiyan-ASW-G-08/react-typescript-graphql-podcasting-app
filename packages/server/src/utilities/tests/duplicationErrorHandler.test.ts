import { Document } from 'mongoose';
import duplicationErrorHandler from '@customUtilities/duplicationErrorHandler';

describe('duplicationErrorHandler', (): void => {
  afterEach(() => jest.clearAllMocks());
  afterAll(() => jest.restoreAllMocks());
  it('should call next with UserInputError with  an array of validation errors', (): void => {
    // expect.assertions(2);
    const nextMock = jest.fn();
    const value = 'TestUser';
    const path = 'username';
    const mongooseDuplicationError = {
      errors: {
        username: {
          value,
          path,
        },
      },
    };
    const documentMock = ({} as unknown) as Document;
    expect(() =>
      duplicationErrorHandler(mongooseDuplicationError, documentMock, nextMock),
    ).toThrowErrorMatchingSnapshot();
  });
});
