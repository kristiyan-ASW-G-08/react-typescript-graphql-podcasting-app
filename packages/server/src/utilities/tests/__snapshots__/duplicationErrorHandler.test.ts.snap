// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`duplicationErrorHandler should call next with UserInputError with  an array of validation errors 1`] = `"Validation Error"`;
